ports/nrf/common-hal/digitalio/DigitalInOut.c

```diff
@@ -47,66 +50,56 @@ void common_hal_digitalio_digitalinout_deinit(digitalio_digitalinout_obj_t *self
     if (common_hal_digitalio_digitalinout_deinited(self)) {
         return;
     }
+    nrf_gpio_cfg_default(self->pin->number);
 
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-    nrf_gpio_cfg_default(pin);
-
+    reset_pin_number(self->pin->number);
     self->pin = mp_const_none;
 }
 
 void common_hal_digitalio_digitalinout_switch_to_input(
         digitalio_digitalinout_obj_t *self, digitalio_pull_t pull) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-
-    nrf_gpio_cfg_input(pin, NRF_GPIO_PIN_NOPULL);
-
+    self->output = false;
+    nrf_gpio_cfg_input(self->pin->number, NRF_GPIO_PIN_NOPULL);
     common_hal_digitalio_digitalinout_set_pull(self, pull);
 }
 
 void common_hal_digitalio_digitalinout_switch_to_output(
         digitalio_digitalinout_obj_t *self, bool value,
         digitalio_drive_mode_t drive_mode) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-
+    self->output = true;
     self->open_drain = (drive_mode == DRIVE_MODE_OPEN_DRAIN);
 
-    nrf_gpio_cfg_input(pin, NRF_GPIO_PIN_NOPULL);
+    nrf_gpio_cfg_input(self->pin->number, NRF_GPIO_PIN_NOPULL);
 
     common_hal_digitalio_digitalinout_set_value(self, value);
 }
 
 digitalio_direction_t common_hal_digitalio_digitalinout_get_direction(
         digitalio_digitalinout_obj_t *self) {
-    const uint32_t pin = NRF_GPIO_PIN_MAP(self->pin->port, self->pin->pin);
-
-    return (nrf_gpio_pin_dir_get(pin) == NRF_GPIO_PIN_DIR_OUTPUT) ? DIRECTION_OUTPUT : DIRECTION_INPUT;
+    return self->output ? DIRECTION_OUTPUT : DIRECTION_INPUT;
```

Comment:
Why change to a variable that has a (low) potential to get out of sync with the HW instead of reading the actual HW state like before?

Reply:
Open-drain outputs set to `.value = True` (HIGH) are simulated by switching the GPIO to being an input with no pullups in `common_hal_digitalio_digitalinout_set_value()`, so the direction is not a reliable indicator.

However, we could use `PIN_CNF` `DRIVE` `SOD1` mode to not need this trick. The nRF GPIO periphal has more modes than the SAMD chips. I was going to revisit this code later amd also turn on 'High drive' for everything, which we also do in `atmel-samd`.